3_javascript/javascript1/week3/script.js

```diff
@@ -0,0 +1,234 @@
+    console.error('Error: ID must be a number.');
+    return;
+  }
+
+  for (const note of notes) {
+    if (note.id === id) {
+      return note;
+    }
+  }
+
+  console.error(`Error: No note found with ID ${id}`);
+}
+
+function logOutNotesFormatted() {
+  for (const note of notes) {
+    console.log(`The note with id: ${note.id}, has the following note text: "${note.content}"`);
+  }
+}
+
+// Unique Feature: Add the ability to delete a note by its ID.
+function deleteNote(id) {
+  if (typeof id !== 'number') {
+    console.error('Error: ID must be a number.');
+    return;
+  }
+
+  const indexToDelete = notes.findIndex((note) => note.id === id);
+
+  if (indexToDelete !== -1) {
+    notes.splice(indexToDelete, 1);
+    console.log(`Note with ID ${id} has been deleted.`);
+  } else {
+    console.error(`Error: No note found with ID ${id}`);
+  }
+}
+
+// Example usage:
+saveNote("Pick up groceries", 1);
+saveNote("Do laundry", 2);
+
+console.log(notes);
+
+const firstNote = getNote(1);
+console.log(firstNote);
+
+logOutNotesFormatted();
+
+deleteNote(2);
+console.log(notes);
+
+
+//CactusIO-interactive (Smart phone usage app) 
+// Create a variable to store activities
+const activities = [];
+
+// Function to add an activity
+function addActivity(date, activity, duration) {
+  // Automatically figure out the date if not provided
+  if (!date) {
+    const now = new Date();
+    date = `${now.getDate()}/${now.getMonth() + 1}-${now.getFullYear()}`;
+  }
+
+  // Create a new activity object
+  const newActivity = { date, activity, duration };
+
+  // Push the activity object into the activities array
+  activities.push(newActivity);
+}
+
+// Function to show status
+function showStatus(activities, limit) {
+  if (activities.length === 0) {
+    console.log("Add some activities before calling showStatus.");
+    return;
+  }
+
+  const totalDuration = activities.reduce((acc, activity) => acc + activity.duration, 0);
+  const activityCount = activities.length;
+
+  let statusMessage = `You have added ${activityCount} activities. They amount to ${totalDuration} min. of usage.`;
+
+  if (limit && totalDuration > limit) {
+    statusMessage += " You have reached your limit, no more smartphoning for you!";
+  }
+
+  console.log(statusMessage);
+}
+
+// Function to find the most time-consuming activity
+function findMostTimeConsumingActivity(activities) {
+  if (activities.length === 0) {
+    console.log("No activities to analyze.");
+    return;
+  }
+
+  const mostTimeConsumingActivity = activities.reduce((max, activity) =>
+    activity.duration > max.duration ? activity : max
+  );
+
+  console.log(`The most time-consuming activity is "${mostTimeConsumingActivity.activity}" with ${mostTimeConsumingActivity.duration} min.`);
+}
+
+// Example usage:
+addActivity("23/7-18", "Youtube", 30);
+addActivity("23/7-18", "Facebook", 20);
+addActivity("24/7-18", "News", 28);
+
+showStatus(activities, 60);
+findMostTimeConsumingActivity(activities);
```

Comment:
🌟 First and foremost, your code is neat, organized, and quite intuitive to follow. This shows that you're considering readability, which is an essential trait for every developer.
🌟 The use of the `Date` object to provide a default date when not supplied is a nice touch. It's user-friendly and saves the user the extra effort.
🌟 The use of the `reduce()` method for calculating the `totalDuration` and finding the `mostTimeConsumingActivity` demonstrates a good understanding of array methods in JavaScript.

✅ Suggestions and Improvements:

- **Error Handling:** Consider adding some checks to ensure that activity and duration are provided and are of the right type (string for activity, number for duration) to avoid unexpected behaviors.

- **Code Duplication:** Your `showStatus` and `findMostTimeConsumingActivity` functions both have a check for an empty activities array. Perhaps you can have a helper function that can check this, reducing code repetition.